pkg/stanza/fileconsumer/internal/fileset/fileset.go

```diff
@@ -0,0 +1,77 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package tracker
+
+import (
+	"errors"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fingerprint"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/reader"
+)
+
+type fileset struct {
+	readers []*reader.Reader
+}
+
+func New(cap int) *fileset {
+	return &fileset{readers: make([]*reader.Reader, 0, cap)}
+}
+
+func (set *fileset) Len() int {
+	return len(set.readers)
+}
+
+func (set *fileset) Get() []*reader.Reader {
+	return set.readers
+}
+
+func (set *fileset) Pop() (*reader.Reader, error) {
+	// closes one file and returns metadata wrapped with a nil reader
```

Comment:
Do we need a nil reader? Why not just return the metadata?

Reply:
i kept it because we want the same fileset compatible with m.knownFiles and previousPollFiles (they use reader.Metadata and reader.Reader respectively)

Reply:
Wouldn't something like this work?

```go
func (set *fileset) Match(fp *fingerprint.Fingerprint) *T {
	for idx, m := range set.matchables {
		if fp.StartsWith(m.Fingerprint) {
			set.matchables = append(set.matchables[:idx], set.matchables[idx+1:]...)
			return m
		}
	}
	return nil
}
```

Reply:
I don't think so.
`m.Fingerprint` won't work atleast, not in generics.
We can access fields even though it's present.
I've come across a solution to cater this need.
Will let you know.

Reply:
Alright, thanks. I think adding `Fingerprint()` to the `Matchable` interface would make it accessible too, but curious to see your solution.

Reply:
@djaglowski yeah, that's what I thought of. I also added some basic test cases.
PTAL

Reply:
Thanks. I think this looks good enough to move forward with. We can adjust as necessary in the next PR. Just need to fix CI.

Reply:
@djaglowski looks good now. I guess we can move forward.